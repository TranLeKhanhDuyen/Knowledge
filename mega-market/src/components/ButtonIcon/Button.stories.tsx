import type { Meta, StoryObj } from '@storybook/react'
import { ButtonIcon } from '@components'
import { menuIcon } from '@assets/images'

export default {
  title: 'Components/ButtonIcon',
  component: ButtonIcon,
  tags: ['autodocs'],
  argTypes: {
    variants: { description: 'The variants of button.' },
    size: { description: 'The size of button' },
    additionalClass: { description: 'Add class to the button' },
    children: { description: 'The content of button.' }
  }
} as Meta

type Story = StoryObj<typeof ButtonIcon>

export const Default: Story = {
  args: {
    variants: 'square',
    size: 'md',
    icon: menuIcon,
    width: '26px',
    height: '18px'
  }
}
