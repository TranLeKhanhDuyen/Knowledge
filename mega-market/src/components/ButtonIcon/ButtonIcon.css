.btn {
  border: none;
  cursor: pointer;
  background-color: transparent;
}

.btn-square {
  background: var(--tertiary-color);
  border: 1px solid transparent;
  border-radius: 10px;
  box-shadow: none;
}

.btn-square:hover {
  border-color: var(--primary-color);
  box-shadow: 0 0 8px rgba(0, 0, 0, 0.2);
}

.btn-circle {
  background: var(--tertiary-color);
  border: 1px solid transparent;
  border-radius: 50%;
}

.btn-circle:hover {
  border: 1px solid var(--primary-color);
}

.btn-sm {
  padding: 14px 12px;
}

.btn-md {
  padding: 31px;
}
