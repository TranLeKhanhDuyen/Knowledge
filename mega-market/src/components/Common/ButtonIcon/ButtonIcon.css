.btn {
  border: none;
  cursor: pointer;
  background-color: transparent;
}

.btn-square {
  background: var(--tertiary-color);
  border: 1px solid transparent;
  border-radius: 10px;
  box-shadow: none;
  width: fit-content;
  height: fit-content;
}

.btn-circle {
  background: var(--tertiary-color);
  border: 1px solid transparent;
  border-radius: 50%;
  padding: 15px;
}

.btn-circle:hover {
  border: 1px solid var(--primary-color);
}

.btn-sm {
  padding: 11px;
}

.btn-md {
  padding: 31px;
}
