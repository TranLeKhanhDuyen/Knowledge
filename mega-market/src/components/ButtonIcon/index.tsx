import '@components/ButtonIcon/ButtonIcon.css'

type TButtonIconVariant = 'square' | 'circle'
type TButtonSize = 'sm' | 'md' | 'lg'

export interface IButtonIconProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variants: TButtonIconVariant
  size?: TButtonSize
  icon: string
  alt: string
  additionalClass?: string
  width?: string
  height?: string
}

const ButtonIcon = ({
  variants = 'square',
  size = 'md',
  icon,
  width,
  height,
  alt = 'icon',
  additionalClass,
  onClick,
  ...props
}: IButtonIconProps) => {
  return (
    <button
      className={`btn btn-${variants} btn-${size} ${additionalClass || ''}`}
      onClick={onClick}
      {...props}
    >
      {icon && (
        <img
          className='btn-icon'
          alt={`Image depicts ${alt}`}
          src={icon}
          width={width}
          height={height}
        />
      )}
    </button>
  )
}

export default ButtonIcon
