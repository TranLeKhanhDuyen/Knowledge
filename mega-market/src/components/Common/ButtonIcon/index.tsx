import IconSvg, { IconSvgProps } from '@components/Common/IconSvg'
import './ButtonIcon.css'

type TButtonIconVariant = 'square' | 'circle'
type TButtonSize = 'sm' | 'md' | 'lg'

export interface IButtonIconProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variants?: TButtonIconVariant
  size?: TButtonSize
  icon: IconSvgProps['name']
  alt?: string
  additionalClass?: string
  onClick?: () => void
}

const ButtonIcon = ({
  variants = 'square',
  size = 'sm',
  icon,
  alt = 'icon',
  additionalClass = '',
  onClick,
  ...props
}: IButtonIconProps) => {
  return (
    <button
      className={`btn btn-${variants} btn-${size} ${additionalClass}`}
      onClick={onClick}
      {...props}
    >
      <IconSvg name={icon} />
    </button>
  )
}

export default ButtonIcon
