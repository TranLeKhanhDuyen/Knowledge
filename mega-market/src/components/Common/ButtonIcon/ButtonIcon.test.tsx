import { render, fireEvent } from '@testing-library/react'
import ButtonIcon, { IButtonIconProps } from '.'

// Mock IconSvg component
jest.mock('@components/Common/IconSvg', () => ({ name }: { name: string }) => (
  <svg data-testid={`icon-${name}`} />
))

describe('ButtonIcon', () => {
  const renderButtonIcon = (props: Partial<IButtonIconProps> = {}) => {
    const defaultProps: IButtonIconProps = {
      variants: 'square',
      size: 'sm',
      icon: 'search',
      alt: 'icon',
      additionalClass: '',
      onClick: jest.fn()
    }
    return render(<ButtonIcon {...defaultProps} {...props} />)
  }

  test('renders with the correct variant class', () => {
    const { container } = renderButtonIcon({ variants: 'circle' })
    const buttonElement = container.querySelector('button')
    expect(buttonElement).toHaveClass('btn btn-circle')
  })

  test('renders with the correct size class', () => {
    const { container } = renderButtonIcon({ size: 'lg' })
    const buttonElement = container.querySelector('button')
    expect(buttonElement).toHaveClass('btn btn-square btn-lg')
  })

  test('renders the correct icon', () => {
    const { getByTestId } = renderButtonIcon({ icon: 'user' })
    expect(getByTestId('icon-user')).toBeInTheDocument()
  })

  test('applies additional class names', () => {
    const { container } = renderButtonIcon({ additionalClass: 'custom-class' })
    const buttonElement = container.querySelector('button')
    expect(buttonElement).toHaveClass('custom-class')
  })

  test('handles the onClick event', () => {
    const onClickMock = jest.fn()
    const { getByRole } = renderButtonIcon({ onClick: onClickMock })
    const buttonElement = getByRole('button')
    fireEvent.click(buttonElement)
    expect(onClickMock).toHaveBeenCalledTimes(1)
  })

  test('matches snapshot', () => {
    const { asFragment } = renderButtonIcon()
    expect(asFragment()).toMatchSnapshot()
  })
})
