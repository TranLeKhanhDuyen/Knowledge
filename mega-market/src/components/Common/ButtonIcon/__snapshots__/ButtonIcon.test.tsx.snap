// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonIcon matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-square btn-sm "
  >
    <svg
      data-testid="icon-search"
    />
  </button>
</DocumentFragment>
`;
